authentication/pipeline/user.py

```diff
@@ -17,3 +24,15 @@ def forbid_hijack(
         msg = "You are hijacking another user, don't try to login again"
         raise AuthException(msg)
     return {}
+
+
+def create_new_user_favorites_list(**kwargs):
+    """
+    Create a favorites UserList for new users.
+    """
+    if kwargs.get("is_new"):
+        pm = get_plugin_manager()
+        hook = pm.hook
+        hook.create_favorites_list(
+            user=User.objects.get(username=kwargs["details"]["username"])
+        )
```

Comment:
So I think the way you want to do this a bit differently because it's unlikely there's going to be multiple ways to create a favorites list or multiple things you do in order to create it. More likely we have a discrete event (in this case a user has been created) and we want to allow arbitrary code to hook into that.

So we'd want the plugin be more abstract and also to define a spec for it:

```python
hookimpl = pluggy.HookimplMarker(APP_NAME)
hookspec = pluggy.HookspecMarker(APP_NAME)


# class is used here just to namespace the spec
class AuthenticationHooks:
    @hookspec
    def user_created(self, user: User): pass
```

Elsewhere:

```python
class FavoritesListPlugin:
    @hookimpl
    def user_created(self, user: User): 
        UserList.objects.get_or_create(
            author=user, title=FAVORITES_TITLE, defaults={"description": "My Favorites"}
        )
```

Then you'd update your pipeline step here to be a more generally named one and over time we can have other kinds of code hook into this same point.

Also, `user` should be available as a kwarg to this function so you shouldn't need to refetch.

Reply:
Thanks, do the app/file locations make sense?  For instance, should the `hooks.py` file containing `hookimpl`, `def get_plugin_manager`, `class AuthenticationHooks` etc be located where it is in the `authentication` app, with `APP_NAME=authentication`, or in the main project app instead (still `open_discussions`, should probably be renamed at some point to `mitopen`)?

Also:
```
class FavoritesListPlugin:
    @hookimpl
    etc
```

is located in `learning_resources/plugins.py` and contains an import for hookimpl from the `authentication` app (which seems to somewhat defeat the purpose of keeping the apps as separate as possible).  

```
from authentication.hooks import hookimpl
```

Is that okay? Or should I redefine it again instead?

```
hookimpl = HookimplMarker("authentication")
```

Reply:
The import should be fine

Reply:
Actually, what do you think about having a convention where we import and assign `hookimpl` to the `AppConfig` and then access it like this:

```python
from django.apps import apps
hookimpl = apps.get_app_config("authentication").hookimpl
```

Reply:
will give that a shot, thanks